pkg/argo_client/manifests.go

```diff
@@ -1,116 +1,470 @@
+		AppLabelKey:        argoSettings.AppLabelKey,
+		AppName:            app.Name,
+		Namespace:          app.Spec.Destination.Namespace,
+		ApplicationSource:  &source,
+		Repos:              permittedHelmRepos,
+		KustomizeOptions:   argoSettings.KustomizeOptions,
+		KubeVersion:        cluster.Info.ServerVersion,
+		ApiVersions:        cluster.Info.APIVersions,
+		HelmRepoCreds:      permittedHelmCredentials,
+		HelmOptions:        helmOptions,
+		TrackingMethod:     argoSettings.TrackingMethod,
+		EnabledSourceTypes: enabledSourceTypes,
+		ProjectName:        proj.Name,
+		ProjectSourceRepos: proj.Spec.SourceRepos,
+		HasMultipleSources: app.Spec.HasMultipleSources(),
+		RefSources:         refSources,
+	}
+
+	log.Info().Msg("generating manifest with files")
+	stream, err := a.repoClient.GenerateManifestWithFiles(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get manifests with files")
+	}
+
+	log.Info().Msg("sending request")
+	if err := stream.Send(&repoapiclient.ManifestRequestWithFiles{
+		Part: &repoapiclient.ManifestRequestWithFiles_Request{
+			Request: &q,
+		},
+	}); err != nil {
+		return nil, errors.Wrap(err, "failed to send request")
+	}
+
+	log.Info().Msg("sending metadata")
+	if err := stream.Send(&repoapiclient.ManifestRequestWithFiles{
+		Part: &repoapiclient.ManifestRequestWithFiles_Metadata{
+			Metadata: &repoapiclient.ManifestFileMetadata{
+				Checksum: checksum,
+			},
 		},
-		true,
-		new(git.NoopCredsStore),
-		resource.MustParse("0"),
-		nil,
-	)
+	}); err != nil {
+		return nil, errors.Wrap(err, "failed to send metadata")
+	}
+
+	log.Info().Msg("sending file")
+	err = sendFile(ctx, stream, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send manifest stream file: %w", err)
+	}
+
+	log.Info().Msg("receiving repsonse")
+	response, err := stream.CloseAndRecv()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get response")
+	}
+
+	log.Info().Msg("done!")
+	return response.Manifests, nil
 }
 
-func ConvertJsonToYamlManifests(jsonManifests []string) []string {
-	var manifests []string
-	for _, manifest := range jsonManifests {
-		ret, err := yaml.JSONToYAML([]byte(manifest))
+func copyFile(srcpath, dstpath string) error {
+	dstdir := filepath.Dir(dstpath)
+	if err := os.MkdirAll(dstdir, 0o777); err != nil {
+		return errors.Wrap(err, "failed to make directories")
+	}
+
+	r, err := os.Open(srcpath)
+	if err != nil {
+		return err
+	}
+	defer r.Close() // ignore error: file was opened read-only.
+
+	w, err := os.Create(dstpath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// Report the error, if any, from Close, but do so
+		// only if there isn't already an outgoing error.
+		if c := w.Close(); err == nil {
+			err = c
+		}
+	}()
+
+	_, err = io.Copy(w, r)
+	return err
+}
+
+func packageApp(ctx context.Context, source v1alpha1.ApplicationSource, refs []v1alpha1.ApplicationSource, repo *git.Repo, getRepo getRepo) (string, error) {
```

Comment:
Can we unittest this? It's critical part of the application that don't involve external calls

Reply:
This test is gonna suck, but you're not wrong, I'll figure it out.